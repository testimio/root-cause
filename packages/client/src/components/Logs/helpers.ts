import { NormalizedLog, CallFrame, LogLevel } from './interfaces';
import { ConsoleMessage, ConsoleException } from '@testim/root-cause-types';
import { StackUtils } from '../../utils/StackUtils';

const stackUtils = new StackUtils({ cwd: process.cwd() });

export function normalizeError(error: ConsoleException): NormalizedLog {
  const stackTrace = getCallFramesFromError(error);
  return {
    level: 'error',
    message: error.message,
    stackTrace: stackTrace,
    when: new Date(error.timestamp),
    text: error.message,
    location: stackTrace ? `${stackTrace[0].lineNumber}:${stackTrace[0].columnNumber}` : '',
    url: stackTrace ? stackTrace[0].url : '',
    type: 'error',
    command: 'error',
    source: 'javascript',
  };
}

function getCallFramesFromError(error: ConsoleException): CallFrame[] | undefined {
  if (!error.stack) {
    return;
  }

  // remove first error line
  const stackTraceLines = error.stack.split('\n').slice(1);

  const frames = stackTraceLines
    ?.map((line) => stackUtils.parseLine(line))
    .filter((frame): frame is NonNullable<typeof frame> => {
      return frame !== null;
    });

  return frames.map((frame) => ({
    functionName: frame.function || '',
    scriptId: '0',
    url: frame.file || '',
    lineNumber: frame.line ?? 0,
    columnNumber: frame.column ?? 0,
  }));
}

export function normalizeConsoleEntry(consoleEntry: ConsoleMessage): NormalizedLog {
  return {
    // Bummer, but most simple solution
    level: consoleEntry.level as LogLevel,
    message: consoleEntry.text,
    when: new Date(consoleEntry.timestamp),
    text: consoleEntry.text,
    location: `${consoleEntry.line}:${consoleEntry.column}`,
    url: consoleEntry.url,
    type: 'logEntry',
    // Bummer, but most simple solution
    command: consoleEntry.level as LogLevel,
    source: 'javascript',
  };
}
