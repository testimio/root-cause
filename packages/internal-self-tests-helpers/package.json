{
  "name": "@testim/internal-self-tests-helpers",
  "version": "0.0.1",
  "private": true,
  "main": "index",
  "devDependencies": {
    "@types/har-format": "^1.2.4",
    "@types/jest": "^26.0.13",
    "@types/node": "^12",
    "escape-string-regexp": "^4.0.0",
    "pretty-format": "^26.4.0",
    "strip-ansi": "^6.0.0",
    "typescript": "^3.9.7"
  }
}
