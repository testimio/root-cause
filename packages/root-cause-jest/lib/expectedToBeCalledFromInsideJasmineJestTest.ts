import type { JasmineCurrentTestInfo, FailedExpectations } from './interfaces';
import type { TestEndStatus } from '@testim/root-cause-types';
/**
 * Helper that masks access to the jasmine global
 */
export function getJasmineCurrentTest(): JasmineCurrentTestInfo {
  // @ts-ignore
  if (typeof jasmine === 'undefined') {
    throw new Error('global jasmine is missing');
  }

  // @ts-ignore
  if (typeof jasmine.currentTest === 'undefined') {
    throw new Error('global jasmine.currentTest is missing');
  }

  // @ts-expect-error
  return jasmine.currentTest;
}

/**
 * Helper that masks access to the jasmine global
 */
export function getEndStatusFromJasmineJest(): TestEndStatus<unknown, FailedExpectations> {
  const t = getJasmineCurrentTest();
  const testEndStatus =
    t.failedExpectations.length > 0
      ? ({
          success: false,
          error: t.failedExpectations[0],
        } as const)
      : ({
          success: true,
          data: t.passedExpectations,
        } as const);

  return testEndStatus;
}
