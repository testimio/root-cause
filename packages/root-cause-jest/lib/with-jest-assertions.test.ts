/* eslint-disable import/no-extraneous-dependencies */
import { attach, utils, updateHistoryFromRootCauseResultsOnly } from '@testim/root-cause-core';
import { getCleanAllPathsPrettyFormatPlugin, getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin } from '@testim/internal-self-tests-helpers';
import puppeteer from 'puppeteer';
import fs from 'fs-extra';
import path from 'path';
import { registerJasmineReporterToGlobal, makeHookExpect } from './helpers';
import { getJasmineCurrentTest } from './expectedToBeCalledFromInsideJasmineJestTest';
import { TestResultFile } from '@testim/root-cause-types';

const { testUniqueIdentifierFromStartParams, jsonReduceNoiseReviver } = utils;

describe('Sanity integration test', () => {
    jest.setTimeout(30_000);
    expect.addSnapshotSerializer(getCleanAllPathsPrettyFormatPlugin(process.cwd()));
    expect.addSnapshotSerializer(getCleanProcessTicksAndRejectionsStackFramePrettyFormatPlugin());

    let browser: puppeteer.Browser;
    let browserContext: puppeteer.BrowserContext;
    let page: puppeteer.Page;

    beforeAll(async () => {
        registerJasmineReporterToGlobal();

        // https://github.com/puppeteer/puppeteer/blob/master/docs/troubleshooting.md#setting-up-chrome-linux-sandbox
        // Not optimal, but didn't work on circle on circleci/node:12.17-stretch-browsers without it
        // Need to revisit
        browser = await puppeteer.launch({ args: ['--no-sandbox', '--disable-setuid-sandbox'] });
    });

    afterAll(async () => {
        await browser.close();
    });

    beforeEach(async () => {
        browserContext = await browser.createIncognitoBrowserContext();
        page = await browserContext.newPage();
    });

    afterEach(async () => {
        await page.close();
        await browserContext.close();
    });

    test('with jest assertions', async () => {
        const currentTest = getJasmineCurrentTest();

        const localRunId = new Date().toString();

        const startTestParams = {
            runId: localRunId,
            projectRoot: process.cwd(),
            fullName: currentTest.fullName,
            description: currentTest.description,
            fullSuitePath: currentTest.testPath,
        };

        const mockedDateConstructor: typeof Date = Object.create(Date);

        let nowCallsCounter = 1;
        mockedDateConstructor.now = function mockedNow() {
            return nowCallsCounter++;
        };

        const attachController = await attach({
            page,
            startTestParams,

        }, mockedDateConstructor);

        const {
            page: playedPage, endTest,
        } = attachController;

        const unhookExpect = makeHookExpect(attachController);

        await playedPage.goto('http://jsbin.testim.io/ner/1');
        await playedPage.click('#test');

        await playedPage.click('#test');
        await playedPage.click('#test');

        await playedPage.click('#test');
        const counterEl = await playedPage.$('#counter');

        expect(counterEl).not.toBe(null);

        utils.assertNotNullOrUndefined(counterEl);

        const text = await (await counterEl.getProperty('innerText')).jsonValue();

        expect(text).toBe('4');

        await playedPage.click('#reset');

        const text2 = await (await counterEl.getProperty('innerText')).jsonValue();

        expect(text2).toBe('0');

        const allButtons = await playedPage.$$('button');

        await allButtons[0].click();

        const text3 = await (await counterEl.getProperty('innerText')).jsonValue();

        expect(text3).toBe('1');

        expect(allButtons).toHaveLength(2);

        await endTest({
            success: false,
            error: { message: 'mocked error test failed' },
        });

        updateHistoryFromRootCauseResultsOnly(localRunId);

        unhookExpect();

        const expectedResultsPath = path.resolve(process.cwd(), '.root-cause', 'runs', localRunId, testUniqueIdentifierFromStartParams(startTestParams));
        const testResults: TestResultFile = JSON.parse(await fs.readFile(path.resolve(expectedResultsPath, 'results.json'), 'utf-8'), jsonReduceNoiseReviver);

        expect(testResults).toMatchSnapshot();

        for (const stepResult of testResults.steps) {
            // Do also images blob comparison? TBH
            if (stepResult.screenshot) {
                // eslint-disable-next-line no-await-in-loop
                expect(fs.pathExists(path.resolve(expectedResultsPath, stepResult.screenshot))).resolves.toBe(true);
            }
        }
    });
});
