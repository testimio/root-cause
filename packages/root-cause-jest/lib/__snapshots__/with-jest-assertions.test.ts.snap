// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity integration test with jest assertions 1`] = `
Object {
  "metadata": Object {
    "branchInfo": "noise_removed:branchInfo",
    "endedTimestamp": 32,
    "fileName": "noise_removed/packages/root-cause-jest/lib/with-jest-assertions.test.ts",
    "hasNetworkLogs": true,
    "systemInfo": Object {
      "automationFramework": "puppeteer",
      "browser": "chromium",
      "browserPlatform": "noise_removed:browserPlatform",
      "browserVersion": "HeadlessChrome/86.0.4240.0",
      "modelName": "noise_removed:modelName",
      "modelVersion": "noise_removed:modelVersion",
      "pageViewport": Object {
        "height": 600,
        "width": 800,
      },
      "userAgent": "noise_removed:userAgent",
    },
    "testEndStatus": Object {
      "codeLocationDetails": Object {
        "callstack": Array [
          Object {
            "column": 29,
            "file": "../root-cause-core/lib/assortedHooks.ts",
            "function": "testEndHook",
            "line": 118,
          },
          Object {
            "column": 15,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "PuppeteerPageHooker.end",
            "line": 188,
          },
          Object {
            "column": 5,
            "file": "../root-cause-core/lib/attach.ts",
            "function": "endTest",
            "line": 115,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 116,
          },
        ],
        "codeLines": Array [
          "",
          "    expect(allButtons).toHaveLength(2);",
          "",
          "    await endTest({",
          "      success: false,",
          "      error: { message: 'mocked error test failed' },",
          "    });",
        ],
        "column": 5,
        "fromRowNumber": 113,
        "row": 116,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 119,
      },
      "error": Object {
        "message": "mocked error test failed",
        "name": "",
        "stack": "",
      },
      "success": false,
    },
    "testFullName": "Sanity integration test with jest assertions",
    "testName": "with jest assertions",
    "timestamp": 2,
  },
  "steps": Array [
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 4,
      "fnName": "goto",
      "index": 1,
      "name": "goto \\"http://jsbin.testim.io/ner/1\\"",
      "screenshot": "1.page-goto.jpg",
      "startTimestamp": 3,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_goto",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 83,
          },
        ],
        "codeLines": Array [
          "      startTestParams.projectRoot",
          "    );",
          "",
          "    await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
        ],
        "column": 5,
        "fromRowNumber": 80,
        "row": 83,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 86,
      },
      "text": "http://jsbin.testim.io/ner/1",
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 6,
      "fnName": "click",
      "index": 2,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "2.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 5,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 84,
          },
        ],
        "codeLines": Array [
          "    );",
          "",
          "    await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
          "    await playedPage.click('#test');",
        ],
        "column": 5,
        "fromRowNumber": 81,
        "row": 84,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 87,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 8,
      "fnName": "click",
      "index": 3,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "3.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 7,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 86,
          },
        ],
        "codeLines": Array [
          "    await playedPage.goto('http://jsbin.testim.io/ner/1');",
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
        ],
        "column": 5,
        "fromRowNumber": 83,
        "row": 86,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 89,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 10,
      "fnName": "click",
      "index": 4,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "4.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 9,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 87,
          },
        ],
        "codeLines": Array [
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
          "    const counterEl = await playedPage.$('#counter');",
        ],
        "column": 5,
        "fromRowNumber": 84,
        "row": 87,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 90,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 12,
      "fnName": "click",
      "index": 5,
      "name": "click \\"#test\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "5.page-click.jpg",
      "selector": "#test",
      "startTimestamp": 11,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 89,
          },
        ],
        "codeLines": Array [
          "    await playedPage.click('#test');",
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
          "    const counterEl = await playedPage.$('#counter');",
          "",
          "    expect(counterEl).not.toBe(null);",
        ],
        "column": 5,
        "fromRowNumber": 86,
        "row": 89,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 92,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 14,
      "fnName": "$",
      "index": 6,
      "name": "$ \\"#counter\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "6.page-$.jpg",
      "selector": "#counter",
      "startTimestamp": 13,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 30,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.returningElementHandlesViaPromiseWrappedFunction_Page_$",
            "line": 122,
          },
          Object {
            "column": 23,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 90,
          },
        ],
        "codeLines": Array [
          "    await playedPage.click('#test');",
          "",
          "    await playedPage.click('#test');",
          "    const counterEl = await playedPage.$('#counter');",
          "",
          "    expect(counterEl).not.toBe(null);",
          "",
        ],
        "column": 23,
        "fromRowNumber": 87,
        "row": 90,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 93,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "fnName": "assertion",
      "index": 7,
      "name": "expect(instanceof(ElementHandle)).not.toBe(null)",
      "startTimestamp": 15,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 33,
            "file": "lib/helpers.ts",
            "function": "Object.sync",
            "line": 212,
          },
          Object {
            "column": 29,
            "file": "lib/hookExpect.ts",
            "function": "Proxy.wrappedFunction_not_toBe",
            "line": 112,
          },
          Object {
            "column": 27,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 92,
          },
        ],
        "codeLines": Array [
          "    await playedPage.click('#test');",
          "    const counterEl = await playedPage.$('#counter');",
          "",
          "    expect(counterEl).not.toBe(null);",
          "",
          "    utils.assertNotNullOrUndefined(counterEl);",
          "",
        ],
        "column": 27,
        "fromRowNumber": 89,
        "row": 92,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 95,
      },
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 17,
      "fnName": "getProperty",
      "index": 8,
      "name": "getProperty \\"innerText\\" on \\"#counter\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "8.elementhandle-getProperty.jpg",
      "selector": "#counter",
      "startTimestamp": 16,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_getProperty",
            "line": 156,
          },
          Object {
            "column": 25,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 96,
          },
        ],
        "codeLines": Array [
          "",
          "    utils.assertNotNullOrUndefined(counterEl);",
          "",
          "    const text = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    expect(text).toBe('4');",
          "",
        ],
        "column": 25,
        "fromRowNumber": 93,
        "row": 96,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 99,
      },
      "text": "innerText",
      "unhandledExceptions": Array [],
    },
    Object {
      "fnName": "assertion",
      "index": 9,
      "name": "expect(4).toBe(4)",
      "startTimestamp": 18,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 33,
            "file": "lib/helpers.ts",
            "function": "Object.sync",
            "line": 212,
          },
          Object {
            "column": 29,
            "file": "lib/hookExpect.ts",
            "function": "Proxy.wrappedFunction_root_toBe",
            "line": 112,
          },
          Object {
            "column": 18,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 98,
          },
        ],
        "codeLines": Array [
          "",
          "    const text = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    expect(text).toBe('4');",
          "",
          "    await playedPage.click('#reset');",
          "",
        ],
        "column": 18,
        "fromRowNumber": 95,
        "row": 98,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 101,
      },
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 20,
      "fnName": "click",
      "index": 10,
      "name": "click \\"#reset\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "10.page-click.jpg",
      "selector": "#reset",
      "startTimestamp": 19,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_Page_click",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 100,
          },
        ],
        "codeLines": Array [
          "",
          "    expect(text).toBe('4');",
          "",
          "    await playedPage.click('#reset');",
          "",
          "    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
        ],
        "column": 5,
        "fromRowNumber": 97,
        "row": 100,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 103,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 22,
      "fnName": "getProperty",
      "index": 11,
      "name": "getProperty \\"innerText\\" on \\"#counter\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "11.elementhandle-getProperty.jpg",
      "selector": "#counter",
      "startTimestamp": 21,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_getProperty",
            "line": 156,
          },
          Object {
            "column": 26,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 102,
          },
        ],
        "codeLines": Array [
          "",
          "    await playedPage.click('#reset');",
          "",
          "    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    expect(text2).toBe('0');",
          "",
        ],
        "column": 26,
        "fromRowNumber": 99,
        "row": 102,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 105,
      },
      "text": "innerText",
      "unhandledExceptions": Array [],
    },
    Object {
      "fnName": "assertion",
      "index": 12,
      "name": "expect(0).toBe(0)",
      "startTimestamp": 23,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 33,
            "file": "lib/helpers.ts",
            "function": "Object.sync",
            "line": 212,
          },
          Object {
            "column": 29,
            "file": "lib/hookExpect.ts",
            "function": "Proxy.wrappedFunction_root_toBe",
            "line": 112,
          },
          Object {
            "column": 19,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 104,
          },
        ],
        "codeLines": Array [
          "",
          "    const text2 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    expect(text2).toBe('0');",
          "",
          "    const allButtons = await playedPage.$$('button');",
          "",
        ],
        "column": 19,
        "fromRowNumber": 101,
        "row": 104,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 107,
      },
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 25,
      "fnName": "$$",
      "index": 13,
      "name": "$$ \\"button\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "13.page-$$.jpg",
      "selector": "button",
      "startTimestamp": 24,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 30,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.returningElementHandlesViaPromiseWrappedFunction_Page_$$",
            "line": 122,
          },
          Object {
            "column": 24,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 106,
          },
        ],
        "codeLines": Array [
          "",
          "    expect(text2).toBe('0');",
          "",
          "    const allButtons = await playedPage.$$('button');",
          "",
          "    await allButtons[0].click();",
          "",
        ],
        "column": 24,
        "fromRowNumber": 103,
        "row": 106,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 109,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 27,
      "fnName": "click",
      "index": 14,
      "name": "click \\"button[0]\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "14.elementhandle-click.jpg",
      "selector": "button[0]",
      "startTimestamp": 26,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_click",
            "line": 156,
          },
          Object {
            "column": 5,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 108,
          },
        ],
        "codeLines": Array [
          "",
          "    const allButtons = await playedPage.$$('button');",
          "",
          "    await allButtons[0].click();",
          "",
          "    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
        ],
        "column": 5,
        "fromRowNumber": 105,
        "row": 108,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 111,
      },
      "unhandledExceptions": Array [],
    },
    Object {
      "consoleEntries": Array [],
      "endTimestamp": 29,
      "fnName": "getProperty",
      "index": 15,
      "name": "getProperty \\"innerText\\" on \\"#counter\\"",
      "rect": Object {
        "bottom": "possible noise removed",
        "devicePixelRatio": "possible noise removed",
        "height": "possible noise removed",
        "left": "possible noise removed",
        "right": "possible noise removed",
        "screenHeight": "possible noise removed",
        "screenWidth": "possible noise removed",
        "top": "possible noise removed",
        "width": "possible noise removed",
        "x": "possible noise removed",
        "y": "possible noise removed",
      },
      "screenshot": "15.elementhandle-getProperty.jpg",
      "selector": "#counter",
      "startTimestamp": 28,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 18,
            "file": "../root-cause-core/lib/PuppeteerPageHooker.ts",
            "function": "Proxy.rootCauseWrappedFunction_ElementHandle_getProperty",
            "line": 156,
          },
          Object {
            "column": 26,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 110,
          },
        ],
        "codeLines": Array [
          "",
          "    await allButtons[0].click();",
          "",
          "    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    expect(text3).toBe('1');",
          "",
        ],
        "column": 26,
        "fromRowNumber": 107,
        "row": 110,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 113,
      },
      "text": "innerText",
      "unhandledExceptions": Array [],
    },
    Object {
      "fnName": "assertion",
      "index": 16,
      "name": "expect(1).toBe(1)",
      "startTimestamp": 30,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 33,
            "file": "lib/helpers.ts",
            "function": "Object.sync",
            "line": 212,
          },
          Object {
            "column": 29,
            "file": "lib/hookExpect.ts",
            "function": "Proxy.wrappedFunction_root_toBe",
            "line": 112,
          },
          Object {
            "column": 19,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 112,
          },
        ],
        "codeLines": Array [
          "",
          "    const text3 = await (await counterEl.getProperty('innerText')).jsonValue();",
          "",
          "    expect(text3).toBe('1');",
          "",
          "    expect(allButtons).toHaveLength(2);",
          "",
        ],
        "column": 19,
        "fromRowNumber": 109,
        "row": 112,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 115,
      },
    },
    Object {
      "fnName": "assertion",
      "index": 17,
      "name": "expect(Array(length=2)).toHaveLength(2)",
      "startTimestamp": 31,
      "stepCodeLocation": Object {
        "callstack": Array [
          Object {
            "column": 33,
            "file": "lib/helpers.ts",
            "function": "Object.sync",
            "line": 212,
          },
          Object {
            "column": 29,
            "file": "lib/hookExpect.ts",
            "function": "Proxy.wrappedFunction_root_toHaveLength",
            "line": 112,
          },
          Object {
            "column": 24,
            "file": "lib/with-jest-assertions.test.ts",
            "function": "Object.<anonymous>",
            "line": 114,
          },
        ],
        "codeLines": Array [
          "",
          "    expect(text3).toBe('1');",
          "",
          "    expect(allButtons).toHaveLength(2);",
          "",
          "    await endTest({",
          "      success: false,",
        ],
        "column": 24,
        "fromRowNumber": 111,
        "row": 114,
        "sourceFileRelativePath": "lib/with-jest-assertions.test.ts",
        "toRowNumber": 117,
      },
    },
  ],
}
`;
